### ISSUE TITLE
Tensorflow Docker: Failed to get a convolutional algorithm. This is probably because cuDNN failed to initialize. 

### ISSUE THREAD

**Issue by paviddavid** [tensorflow/tensorflow#27715]
Hey all, 

I saw that there are already some issues regarding` Failed to get convolutional algorithm. This is probably because cuDNN failed to initialize, so try look to see if a warning log message was printed above. ` Probably this error is caused by an incompatibility of cuDNN or so. However, I am working within an Docker image. I run `nvidia-docker run -it -v home_Path:container_path --rm tensorflow/tensorflow:latest-gpu-py3 ` and inside this image i run a Faster R-CNN code. The first two/three times everything worked well but afterwards every time this error is printed and the programm stops running. Even if I restart the container (by exiting and executing the same command another time), this error appear.  

I hope you guys can help me. I am struggling since days with this problem and I have really no clue what's the problem or how to solve it. 


Thanks a lot in advance. 
------------------------------------------------------------------------------------------------------
Output: 


`../Lib/faster_rcnn_config.py:313: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  yaml_cfg = edict(yaml.load(f))
Restoring from Cityscapes.yml file
{'seed': 1234, 'restore_slim_file': '/root/OD/tf-Faster-RCNN/Data/ResNet50/resnet_v1_50.ckpt', 'num_epochs': 5, 'file_epoch': 1, 'run_num': 31, 'restore': False, 'restore_num': 1, 'batch_size': 1, 'data_directory': '/root/OD/tf-Faster-RCNN/Data/Cityscapes_for_FasterRCNN_REDUCED_REDUCED_imagesize/', 'save_directory': '../Logs/', 'gpu': 0, 'vis': False, 'display_step': 1000, 'learning_rate': 0.003, 'model_directory': 'Cityscapes/'}
Using GPU 0
WARNING:tensorflow:From /usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/op_def_library.py:263: colocate_with (from tensorflow.python.framework.ops) is deprecated and will be removed in a future version.
Instructions for updating:
Colocations handled automatically by placer.
Tensor("model/resnet_v1_50/block3/unit_6/bottleneck_v1/Relu:0", shape=(1, ?, ?, 1024), dtype=float32)
conv_1 output: (1, ?, ?, 512)
conv_1 output: (1, ?, ?, 18)
WARNING:tensorflow:From ../Networks/anchor_target_layer.py:40: py_func (from tensorflow.python.ops.script_ops) is deprecated and will be removed in a future version.
Instructions for updating:
tf.py_func is deprecated in TF V2. Instead, use
    tf.py_function, which takes a python function which manipulates tf eager
    tensors instead of numpy arrays. It's easy to convert a tf eager tensor to
    an ndarray (just call tensor.numpy()) but having access to eager tensors
    means 'tf.py_function's can use accelerators such as GPUs as well as
    being differentiable using a gradient tape.
    
conv_1 output: (1, ?, ?, 36)
WARNING:tensorflow:From ../Lib/roi_pool.py:31: div (from tensorflow.python.ops.math_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Deprecated in favor of operator or tf.math.divide.
flat_1 output: (?, 50176)
WARNING:tensorflow:From ../Lib/TensorBase/tensorbase/base.py:274: calling dropout (from tensorflow.python.ops.nn_ops) with keep_prob is deprecated and will be removed in a future version.
Instructions for updating:
Please use 'rate' instead of 'keep_prob'. Rate should be set to 'rate = 1 - keep_prob'.
fc_1 output: (?, 4096)
fc_2 output: (?, 4096)
fc_1 output: (?, 4)
fc_1 output: (?, 16)
Tensor("model_1/resnet_v1_50/block3/unit_6/bottleneck_v1/Relu:0", shape=(1, ?, ?, 1024), dtype=float32)
conv_1 output: (1, ?, ?, 512)
conv_1 output: (1, ?, ?, 18)
conv_1 output: (1, ?, ?, 36)
flat_1 output: (?, 50176)
fc_1 output: (?, 4096)
fc_2 output: (?, 4096)
fc_1 output: (?, 4)
fc_1 output: (?, 16)
WARNING:tensorflow:From /usr/local/lib/python3.5/dist-packages/tensorflow/python/ops/math_ops.py:3066: to_int32 (from tensorflow.python.ops.math_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.cast instead.
/usr/local/lib/python3.5/dist-packages/tensorflow/python/ops/gradients_impl.py:110: UserWarning: Converting sparse IndexedSlices to a dense Tensor of unknown shape. This may consume a large amount of memory.
  "Converting sparse IndexedSlices to a dense Tensor of unknown shape. "
2019-04-10 15:29:35.327116: I tensorflow/compiler/xla/service/service.cc:150] XLA service 0x7d5c080 executing computations on platform CUDA. Devices:
2019-04-10 15:29:35.327156: I tensorflow/compiler/xla/service/service.cc:158]   StreamExecutor device (0): GeForce RTX 2070, Compute Capability 7.5
2019-04-10 15:29:35.329004: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 3700305000 Hz
2019-04-10 15:29:35.329289: I tensorflow/compiler/xla/service/service.cc:150] XLA service 0x7dc5f20 executing computations on platform Host. Devices:
2019-04-10 15:29:35.329312: I tensorflow/compiler/xla/service/service.cc:158]   StreamExecutor device (0): <undefined>, <undefined>
2019-04-10 15:29:35.329518: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1433] Found device 0 with properties: 
name: GeForce RTX 2070 major: 7 minor: 5 memoryClockRate(GHz): 1.62
pciBusID: 0000:04:00.0
totalMemory: 7.76GiB freeMemory: 6.97GiB
2019-04-10 15:29:35.329541: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1512] Adding visible gpu devices: 0
2019-04-10 15:29:35.493916: I tensorflow/core/common_runtime/gpu/gpu_device.cc:984] Device interconnect StreamExecutor with strength 1 edge matrix:
2019-04-10 15:29:35.493961: I tensorflow/core/common_runtime/gpu/gpu_device.cc:990]      0 
2019-04-10 15:29:35.493975: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1003] 0:   N 
2019-04-10 15:29:35.494236: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1115] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 7871 MB memory) -> physical GPU (device: 0, name: GeForce RTX 2070, pci bus id: 0000:04:00.0, compute capability: 7.5)
2019-04-10 15:29:36.902368: E tensorflow/stream_executor/cuda/cuda_driver.cc:806] failed to allocate 7.69G (8253855488 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY: out of memory
Restoring TF-Slim Model.
WARNING:tensorflow:From /usr/local/lib/python3.5/dist-packages/tensorflow/python/training/saver.py:1266: checkpoint_exists (from tensorflow.python.training.checkpoint_management) is deprecated and will be removed in a future version.
Instructions for updating:
Use standard file APIs to check for files with this prefix.
INFO:tensorflow:Restoring parameters from /root/OD/tf-Faster-RCNN/Data/ResNet50/resnet_v1_50.ckpt
2019-04-10 15:29:38.656957: W tensorflow/core/framework/allocator.cc:124] Allocation of 822083584 exceeds 10% of system memory.
2019-04-10 15:29:38.879858: W tensorflow/core/framework/allocator.cc:124] Allocation of 822083584 exceeds 10% of system memory.
2019-04-10 15:29:39.126316: W tensorflow/core/framework/allocator.cc:124] Allocation of 822083584 exceeds 10% of system memory.
2019-04-10 15:29:39.343851: W tensorflow/core/framework/allocator.cc:124] Allocation of 822083584 exceeds 10% of system memory.
2019-04-10 15:29:39.831089: W tensorflow/core/framework/allocator.cc:124] Allocation of 822083584 exceeds 10% of system memory.
Learning Rate: 0.003000
Epochs: 5
None
{'seed': 1234, 'restore_slim_file': '/root/OD/tf-Faster-RCNN/Data/ResNet50/resnet_v1_50.ckpt', 'num_epochs': 5, 'file_epoch': 1, 'run_num': 31, 'restore_directory': '../Logs/Cityscapes/Model1/', 'restore': False, 'restore_num': 1, 'batch_size': 1, 'data_directory': '/root/OD/tf-Faster-RCNN/Data/Cityscapes_for_FasterRCNN_REDUCED_REDUCED_imagesize/', 'save_directory': '../Logs/', 'gpu': 0, 'vis': False, 'display_step': 1000, 'logging_directory': '../Logs/Cityscapes/Model31/', 'learning_rate': 0.003, 'model_directory': 'Cityscapes/'}
Training for 5 epochs
epochs:   0%|                                                                                                                                                                                                                                                | 0/5 [00:00<?, ?it/s2019-04-10 15:29:41.114390: I tensorflow/stream_executor/dso_loader.cc:152] successfully opened CUDA library libcublas.so.10.0 locally                                                                                                                     | 0/2096 [00:00<?, ?it/s]
2019-04-10 15:29:41.140134: E tensorflow/stream_executor/cuda/cuda_blas.cc:510] failed to create cublas handle: CUBLAS_STATUS_NOT_INITIALIZED
2019-04-10 15:29:41.141836: E tensorflow/stream_executor/cuda/cuda_blas.cc:510] failed to create cublas handle: CUBLAS_STATUS_NOT_INITIALIZED
2019-04-10 15:29:41.143191: E tensorflow/stream_executor/cuda/cuda_blas.cc:510] failed to create cublas handle: CUBLAS_STATUS_NOT_INITIALIZED
2019-04-10 15:29:41.144476: E tensorflow/stream_executor/cuda/cuda_blas.cc:510] failed to create cublas handle: CUBLAS_STATUS_NOT_INITIALIZED
2019-04-10 15:29:41.154047: E tensorflow/stream_executor/cuda/cuda_blas.cc:510] failed to create cublas handle: CUBLAS_STATUS_NOT_INITIALIZED
2019-04-10 15:29:41.155242: E tensorflow/stream_executor/cuda/cuda_blas.cc:510] failed to create cublas handle: CUBLAS_STATUS_NOT_INITIALIZED
2019-04-10 15:29:41.157620: E tensorflow/stream_executor/cuda/cuda_blas.cc:510] failed to create cublas handle: CUBLAS_STATUS_NOT_INITIALIZED
2019-04-10 15:29:41.158772: E tensorflow/stream_executor/cuda/cuda_blas.cc:510] failed to create cublas handle: CUBLAS_STATUS_NOT_INITIALIZED
2019-04-10 15:29:41.161637: E tensorflow/stream_executor/cuda/cuda_blas.cc:510] failed to create cublas handle: CUBLAS_STATUS_NOT_INITIALIZED
2019-04-10 15:29:41.162776: E tensorflow/stream_executor/cuda/cuda_blas.cc:510] failed to create cublas handle: CUBLAS_STATUS_NOT_INITIALIZED
2019-04-10 15:29:41.165628: E tensorflow/stream_executor/cuda/cuda_blas.cc:510] failed to create cublas handle: CUBLAS_STATUS_NOT_INITIALIZED
2019-04-10 15:29:41.166792: E tensorflow/stream_executor/cuda/cuda_blas.cc:510] failed to create cublas handle: CUBLAS_STATUS_NOT_INITIALIZED
2019-04-10 15:29:41.195178: E tensorflow/stream_executor/cuda/cuda_dnn.cc:334] Could not create cudnn handle: CUDNN_STATUS_INTERNAL_ERROR
2019-04-10 15:29:41.197679: E tensorflow/stream_executor/cuda/cuda_dnn.cc:334] Could not create cudnn handle: CUDNN_STATUS_INTERNAL_ERROR

Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/client/session.py", line 1334, in _do_call
    return fn(*args)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/client/session.py", line 1319, in _run_fn
    options, feed_dict, fetch_list, target_list, run_metadata)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/client/session.py", line 1407, in _call_tf_sessionrun
    run_metadata)
tensorflow.python.framework.errors_impl.UnknownError: Failed to get convolution algorithm. This is probably because cuDNN failed to initialize, so try looking to see if a warning log message was printed above.
	 [[{{node model/resnet_v1_50/conv1/Conv2D}}]]
	 [[{{node model/roi_proposal/rpn_softmax/transpose_3}}]]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "faster_rcnn_resnet50ish.py", line 300, in <module>
    main()
  File "faster_rcnn_resnet50ish.py", line 293, in main
    model.train()
  File "faster_rcnn_resnet50ish.py", line 169, in train
    summary = self._run_train_iter(feed_dict)
  File "faster_rcnn_resnet50ish.py", line 137, in _run_train_iter
    summary, _ = self.sess.run([self.merged, self.optimizer], feed_dict=feed_dict)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/client/session.py", line 929, in run
    run_metadata_ptr)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/client/session.py", line 1152, in _run
    feed_dict_tensor, options, run_metadata)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/client/session.py", line 1328, in _do_run
    run_metadata)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/client/session.py", line 1348, in _do_call
    raise type(e)(node_def, op, message)
tensorflow.python.framework.errors_impl.UnknownError: Failed to get convolution algorithm. This is probably because cuDNN failed to initialize, so try looking to see if a warning log message was printed above.
	 [[node model/resnet_v1_50/conv1/Conv2D (defined at ../Networks/resnet50_reduced.py:79) ]]
	 [[node model/roi_proposal/rpn_softmax/transpose_3 (defined at ../Lib/rpn_softmax.py:46) ]]

Caused by op 'model/resnet_v1_50/conv1/Conv2D', defined at:
  File "faster_rcnn_resnet50ish.py", line 300, in <module>
    main()
  File "faster_rcnn_resnet50ish.py", line 291, in main
    model = FasterRcnnRes50(flags, dictionary)
  File "faster_rcnn_resnet50ish.py", line 44, in __init__
    super().__init__(flags_input, flags_input['run_num'], vram=cfg.VRAM, restore=flags_input['restore_num'], restore_slim=flags_input['restore_slim_file'])
  File "../Lib/TensorBase/tensorbase/base.py", line 676, in __init__
    self._network()
  File "faster_rcnn_resnet50ish.py", line 82, in _network
    self._faster_rcnn(self.x['TRAIN'], self.gt_boxes['TRAIN'], self.im_dims['TRAIN'], 'TRAIN')
  File "faster_rcnn_resnet50ish.py", line 93, in _faster_rcnn
    feature_maps = resnet50_reduced(x)
  File "../Networks/resnet50_reduced.py", line 79, in resnet50_reduced
    net = resnet_utils.conv2d_same(net, 64, 7, stride=2, scope='conv1')
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/contrib/slim/python/slim/nets/resnet_utils.py", line 146, in conv2d_same
    scope=scope)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/contrib/framework/python/ops/arg_scope.py", line 182, in func_with_args
    return func(*args, **current_args)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/contrib/layers/python/layers/layers.py", line 1155, in convolution2d
    conv_dims=2)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/contrib/framework/python/ops/arg_scope.py", line 182, in func_with_args
    return func(*args, **current_args)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/contrib/layers/python/layers/layers.py", line 1058, in convolution
    outputs = layer.apply(inputs)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/keras/engine/base_layer.py", line 1227, in apply
    return self.__call__(inputs, *args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/layers/base.py", line 530, in __call__
    outputs = super(Layer, self).__call__(inputs, *args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/keras/engine/base_layer.py", line 554, in __call__
    outputs = self.call(inputs, *args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/keras/layers/convolutional.py", line 194, in call
    outputs = self._convolution_op(inputs, self.kernel)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/ops/nn_ops.py", line 966, in __call__
    return self.conv_op(inp, filter)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/ops/nn_ops.py", line 591, in __call__
    return self.call(inp, filter)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/ops/nn_ops.py", line 208, in __call__
    name=self.name)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/ops/gen_nn_ops.py", line 1026, in conv2d
    data_format=data_format, dilations=dilations, name=name)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/op_def_library.py", line 788, in _apply_op_helper
    op_def=op_def)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/util/deprecation.py", line 507, in new_func
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/ops.py", line 3300, in create_op
    op_def=op_def)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/ops.py", line 1801, in __init__
    self._traceback = tf_stack.extract_stack()

UnknownError (see above for traceback): Failed to get convolution algorithm. This is probably because cuDNN failed to initialize, so try looking to see if a warning log message was printed above.
	 [[node model/resnet_v1_50/conv1/Conv2D (defined at ../Networks/resnet50_reduced.py:79) ]]
	 [[node model/roi_proposal/rpn_softmax/transpose_3 (defined at ../Lib/rpn_softmax.py:46) ]]
`

**Comment by bipinkc19** [comment]
If it helps it generally occurs due to unreleased GPU memory form the previous session of training or GPU memory isn't sufficient for the network weights. As your code ran a few times then stopped it is probably the GPU memory is hogged. I would suggest checking the task manager to see GPU consumption before running the code and after. Also, try searching for how to release GPU memory.

**Comment by achandraa** [comment]
Please provide details about what platform you are using (operating system, architecture). Also include your TensorFlow version. Also, did you compile from source or install a binary?

Make sure you also include the exact command if possible to produce the output included in your test case. If you are unclear what to include see the issue template displayed in [the Github new issue template](https://github.com/tensorflow/tensorflow/issues/new).

We ask for this in the issue submission template, because it is really difficult to help without that information. Thanks!

**Comment by paviddavid** [comment]
@bipinkc19 Thanks. How should I release GPU memory in case it is full? I added the code as mentioned here (https://github.com/tensorflow/tensorflow/issues/24828#issuecomment-464910864) and it worked as expected. However, I tried it only once. Is this the solution, even for running the code in an docker container? 

@achandraa I see your point however, I mentioned I work with Docker and therefore, it should be independent from outer resources, or? 
I am working on a Linux machine (16.04) and I pull the latest Tensorflow-Docker image with the mentioned command `nvidia-docker run -it -v home_Path:container_path --rm tensorflow/tensorflow:latest-gpu-py3` - latest TF is 1.13.1 
Everything that I have done happened inside this Nvidia-Docker container.

**Comment by bipinkc19** [comment]
https://github.com/tensorflow/tensorflow/issues/24828#issuecomment-464960819. I think this one is good as it allows growth too. It shouldn't matter if you run it on docker or not. Also are you closing the session after running the code? Generally even when using ```with tf.Session() as sess:``` in jupyter notebooks, the memory isn't released.

**Comment by paviddavid** [comment]
Hello @bipinkc19 No i think i missed to close the session. The next time, this error occurs, I will try to add the line. Thanks a lot!

**Comment by achandraa** [comment]
@paviddavid : Are you still stuck in the issue ? Can you please let us know if the same is resolved ?

**Comment by paviddavid** [comment]
@achandraa By adding the code I mentioned in previous comment helped me to not re-obtained the issue. For the moment, it works.

**Comment by guaguablue** [comment]
> If it helps it generally occurs due to unreleased GPU memory form the previous session of training or GPU memory isn't sufficient for the network weights. As your code ran a few times then stopped it is probably the GPU memory is hogged. I would suggest checking the task manager to see GPU consumption before running the code and after. Also, try searching for how to release GPU memory.

I faced the same issue, and I tried to reboot my mechine, then it works well. However, after I stoped training for some while, the error came again. And I checked the GPU memory, most of it was idle, it was not like there wasn't sufficient memory for the network weights, so I wondered if there were some necessary processes for cudnn was occupied.

**Comment by guaguablue** [comment]
CUDA : 10.0.130
Tensorflow : 1.13.1
Issue is solved by adding the codes below into the file.

config = tf.ConfigProto()
config.gpu_options.allow_growth = True
config.gpu_options.per_process_gpu_memory_fraction=0.5
sess = tf.Session(config=config)

And the reason why the issue came out is that tensorflow will use nearlly all the memory of GPU in default, so, when other processes use some of the memory , it may cause that there is not enough memory for tensorflow to use. In my case, my GPU memory only has 7952M, and build a tensorflow device takes 7438M, only 514M memory left, and 500M memory is easy to run out. To limit memory for tensorflow, I use "config.gpu_options.per_process_gpu_memory_fraction=0.5" allocate 50% of memory to tf, then the tf device will only takes about 4611M memory and the issue is solved.

**Comment by kernelizd** [comment]
I have the same issue. 
Nothing like above solves it.
TF 2.0 alpha0
CUDA : 10.0.130
Cudnn 7.3.1
For some reason, Conv2D doesn't work

**Comment by chris-opendata** [comment]
I am getting the same error for tf.conv2d that never worked. So, it's not related to the memory. My dev env is also a docker (built according to https://www.tensorflow.org/install/docker) as follows:
    Host OS: Ubuntu 18.04
    GPU: RTX 2070
    NVidia driver: 418.56
    cuda ver: 10.1
    Nvidia cuda docker image: nvidia/cuda:latest
    The custom docker built based on tensorflow/tensorflow:latest-gpu-py3-jupyter.
    tensorflow version: 1.13.1

I couldn't find cuDNN library and its header file neither from my host OS env nor from Nvidia cuda docker container and the tensorflow:latest-gpu-py3-jupyter docker container. It seems that the docker images are missing the cudnn installation. Any help will be appreciated.

Chris

**Comment by paviddavid** [comment]
@chris-opendata Sounds weird.. I am working under Ubuntu 16.04 and I used the docker image tensorflow/tensorflow:latest-gpu-py3 and it worked for me as expected. My Cuda is 10.0. Have you tried to verify cudnn in your image? nvcc --version could maybe help
